File: validation/conservative_source_bench.py
"""Time complete source-default flame sequences, with separate physical references."""
import argparse,ast,hashlib,json,os,platform,re,statistics,subprocess,sys,time,tomllib
from pathlib import Path
THREAD_KEYS=["OPENBLAS_NUM_THREADS","OMP_NUM_THREADS","MKL_NUM_THREADS","VECLIB_MAXIMUM_THREADS","JULIA_NUM_THREADS"]
for key in THREAD_KEYS:os.environ[key]="1"
import numpy as np
import cantera as ct
from benchmark_environment import host_metadata,matches_target,loaded_library_paths,verify_numerical_threads

PINNED_COMMIT="726522be4e2a13454d8415b7ef799d621f665cf3"
SOURCE_EXAMPLES={
    "free":("adiabatic_flame","f2a845e0e2b0c06d466eadeba9b1be1d9b8124aadf94e83230527d4124b5840d"),
    "burner":("burner_flame","8c95f440748a2559b4f2a940b6b1864144e0668704a8d894e5923663278632c2"),
    "fixed":("flame_fixed_T","102af72f0349116ce7a1258bb1ce0b06cc04ebe7359358b5c4d635799bbd01be")}
MECHANISM_HASHES={"h2o2":"0efc6c52862741a29e0c29b65d979c7d8cb409db5282bca83b9c5437b3d8c8d4",
    "gri30":"06650b1e0ee0012f6903d5328b1bb218cb6007d07f8ebe375d18f24811039345"}
def digest(path):return hashlib.sha256(Path(path).read_bytes()).hexdigest()

def current_source_hashes(project):
    return {str(path.relative_to(project)):digest(path) for path in sorted((project/"src").rglob("*.jl"))}

def require_unchanged(expected):
    changed=[path for path,value in expected.items() if not Path(path).is_file() or digest(path)!=value]
    if changed:raise RuntimeError(f"benchmark input/source bytes changed: {changed}")

def checked_times(values,repetitions,stages):
    values=np.asarray(values,dtype=float)
    if values.shape!=(repetitions+1,stages) or not np.all(np.isfinite(values)) or not np.all(values>0):
        raise RuntimeError("missing or invalid first/warm stage timings")
    return np.sum(values,axis=1)

def verified_cantera_libraries(build):
    import cantera._cantera as compiled
    paths={path.resolve() for path in loaded_library_paths() if "cantera" in path.name.lower() and path.is_file()}
    paths.add(Path(compiled.__file__).resolve())
    hashes={str(path):digest(path) for path in sorted(paths)}
    if len(hashes)<2 or any(value not in build.get("library_hashes",{}).values() for value in hashes.values()):
        raise RuntimeError(f"loaded Cantera library/build-record mismatch: {hashes}")
    return hashes

def verify_source_inputs(case,source,mechanism,provenance,profile):
    name,expected=SOURCE_EXAMPLES[case]
    if digest(source)!=expected or provenance.get("source_commit")!=PINNED_COMMIT or provenance.get("examples",{}).get(name,{}).get("sha256")!=expected:
        raise ValueError("example source/provenance does not match pinned Cantera")
    if digest(mechanism)!=MECHANISM_HASHES[mechanism.stem]:raise ValueError("mechanism bytes differ from pinned stock source")
    if case=="fixed":
        values={}
        for node in ast.parse(source.read_text()).body:
            if isinstance(node,ast.Assign) and len(node.targets)==1 and isinstance(node.targets[0],ast.Name) and node.targets[0].id in ("zloc","tvalues"):
                values[node.targets[0].id]=np.asarray(ast.literal_eval(node.value.args[0]),dtype=float)
        if not np.array_equal(values.get("zloc"),profile["positions"]) or not np.array_equal(values.get("tvalues"),profile["temperatures"]):
            raise ValueError("prescribed temperature profile differs from exact source values")

def loaded_libraries():
    return [str(path) for path in loaded_library_paths()]


def verify_threads(*,set_accelerate=False):
    result=verify_numerical_threads(set_accelerate=set_accelerate)
    result["environment"]={key:os.environ[key] for key in THREAD_KEYS}
    return result

def cantera_sequence(gas,case,profile,save_profiles=False):
    free=case=="free";fixed=case=="fixed"
    modes=["mass","mass-soret","multi","multi-soret"] if free else ["mole","multi"]
    seconds=[];points=[];snapshots={}
    # Restore the preloaded gas's initial transport manager, corresponding to
    # fresh mechanism loading outside the timed source calculation.
    gas.transport_model="mixture-averaged"
    for stage,mode in enumerate(modes):
        multi=mode.startswith("multi");start=time.perf_counter()
        if stage==0:
            gas.TPX=(300.,ct.one_atm,"H2:1.1,O2:1,AR:5") if free else (373.7,ct.one_atm,"CH4:.65,O2:1,N2:3.76") if fixed else (373.,.05*ct.one_atm,"H2:1.5,O2:1,AR:7")
            flame=ct.FreeFlame(gas,width=.03) if free else ct.BurnerFlame(gas,width=.01 if fixed else .5)
            if not free:flame.burner.mdot=.04 if fixed else .06
            if fixed:
                flame.flame.set_fixed_temp_profile(profile["positions"]/.01,profile["temperatures"])
                flame.energy_enabled=False
        flame.transport_model="multicomponent" if multi else "mixture-averaged"
        flame.flux_gradient_basis="mass" if free else "molar"
        flame.soret_enabled=mode.endswith("soret")
        slope=.06 if free else (.1 if multi else .3) if fixed else .05
        curve=.12 if free else (.2 if multi else 1.) if fixed else .1
        flame.set_refine_criteria(ratio=3.,slope=slope,curve=curve)
        # Preserve the published default tolerances and Jacobian mode here.
        # Refined accuracy-reference solves are entirely outside this runner.
        flame.solve(loglevel=0,auto=stage==0 and not fixed)
        if save_profiles:
            snapshots[mode]=dict(grid=flame.grid.copy(),T=flame.T.copy(),Y=flame.Y.copy(),velocity=flame.velocity.copy(),
                inlet_Y=flame.inlet.Y.copy() if free else flame.burner.Y.copy(),P=[flame.P])
        seconds.append(time.perf_counter()-start);points.append(len(flame.grid))
    return seconds,points,snapshots

def main():
    p=argparse.ArgumentParser(description=__doc__)
    p.add_argument("parameters",type=Path);p.add_argument("references",type=Path);p.add_argument("output",type=Path)
    p.add_argument("case",choices=["free","burner","fixed"])
    p.add_argument("--project",required=True,type=Path);p.add_argument("--julia",required=True)
    p.add_argument("--build-record",required=True,type=Path)
    p.add_argument("--native-record",required=True,type=Path)
    p.add_argument("--source-examples",required=True,type=Path,help="pinned Cantera samples/python/onedim directory")
    p.add_argument("--target",required=True,choices=["wsl","apple-m4"])
    p.add_argument("--reps",type=int,default=9);p.add_argument("--formal",action="store_true")
    p.add_argument("--order",choices=["cantera-first","julia-first"],default="cantera-first")
    a=p.parse_args()
    if a.reps<5 or (a.formal and a.reps<9):p.error("at least five warm repetitions required; formal qualification requires nine")
    if not ct.__version__.startswith("4.0"):p.error("Cantera4 required")
    build=json.loads(a.build_record.read_text())
    if build.get("source",{}).get("commit")!=PINNED_COMMIT:p.error("build record does not identify pinned pristine source")
    native_record=json.loads(a.native_record.read_text())
    if not re.fullmatch(r"[0-9a-f]{40}",native_record.get("commit","")):p.error("native record requires a full checkpoint commit")
    source_hashes=current_source_hashes(a.project)
    if source_hashes!=native_record.get("source_hashes"):p.error("native source files do not match the pinned checkpoint record")
    host=host_metadata()
    if not matches_target(host,a.target):p.error(f"host mismatch: {host}")
    if a.output.exists() and any(a.output.iterdir()):p.error("output directory must be empty to exclude stale artifacts")
    a.output.mkdir(parents=True,exist_ok=True)
    native=a.output/"native";native.mkdir(exist_ok=True)
    cantera=a.output/"cantera";cantera.mkdir(exist_ok=True)
    mechanism=a.parameters/("gri30.yaml" if a.case=="fixed" else "h2o2.yaml")
    gas=ct.Solution(str(mechanism))
    if (gas.n_species,gas.n_reactions)!=((53,325) if a.case=="fixed" else (10,29)):p.error("exact stock mechanism required")
    profile=np.load(a.parameters/"fixed-profile.npz") if a.case=="fixed" else None
    provenance=json.loads((a.parameters/"provenance.json").read_text())
    example=a.source_examples/(SOURCE_EXAMPLES[a.case][0]+".py")
    verify_source_inputs(a.case,example,mechanism,provenance,profile)
    threads=verify_threads(set_accelerate=True)
    libraries_before=verified_cantera_libraries(build)
    driver_paths=[Path(__file__),Path(__file__).with_suffix(".jl"),*[Path(__file__).with_name(name) for name in
        ("conservative_source_accuracy.py","flame_benchmarks.py","benchmark_environment.py","numerical_threads.jl")]]
    example_paths=[a.project/"example/flames"/name for name in
        ("source_flame_sequence.jl","adiabatic_flame.jl","burner_flame.jl","flame_fixed_T.jl")]
    input_paths=[*driver_paths,*example_paths,example,mechanism,Path(str(mechanism)+".npz"),Path(str(mechanism)+".multicomponent.npz"),
        a.parameters/"provenance.json",a.build_record,a.native_record,a.project/"Project.toml",a.project/"Manifest.toml",
        *sorted(a.references.glob(a.case+"-*.npz"))]
    if a.case=="fixed":input_paths.append(a.parameters/"fixed-profile.npz")
    input_hashes={str(path.resolve()):digest(path) for path in input_paths}
    report=dict(case=a.case,formal=a.formal,passed=False,performance_pass=False,date=time.strftime("%Y-%m-%d %H:%M:%S %z"),host=host,target=a.target,
        scope="Sum of construction/initialization/adaptive-solve stages and required numerical profiles in the complete published transport sequence. Mechanism/sidecar loading, validation and file output excluded. No refined-reference solve is timed.",
        compilation_scope="Julia runtime startup and using/imports are outside timers. Specialization of run_source_sequence before entry to its internal stage timers is also excluded; the first measured sequence is not whole-program cold latency. Only JIT triggered after a stage timer starts can enter its measurement. First repetition is recorded separately and excluded from warm medians.",
        timing_order=a.order,threads=threads,cantera_version=ct.__version__,cantera_build_record=build,
        cantera_build_record_sha256=digest(a.build_record),mechanism_sha256=digest(mechanism),
        sidecar_sha256=digest(str(mechanism)+".npz"),multicomponent_sha256=digest(str(mechanism)+".multicomponent.npz"),
        fixed_temperature_profile_sha256=digest(a.parameters/"fixed-profile.npz") if a.case=="fixed" else None,
        parameter_provenance=provenance,original_source_example=dict(path=str(example),sha256=digest(example)),
        reference_hashes={path.name:digest(path) for path in sorted(a.references.glob(a.case+"-*.npz"))},
        benchmark_driver_hashes={path.name:digest(path) for path in driver_paths},input_hashes=input_hashes,
        native_example_hashes={path.name:digest(path) for path in example_paths},
        actual_loaded_cantera_hashes_before=libraries_before,python_executable_sha256=digest(sys.executable),
        native_source_hashes=source_hashes,native_commit=native_record.get("commit"),native_record_sha256=digest(a.native_record))
    def run_ct():
        times=[];nodes=[];baseline=None;replay=[];thread_checks=[]
        for repetition in range(a.reps+1):
            elapsed,points,snapshots=cantera_sequence(gas,a.case,profile,True)
            thread_checks.append(verify_threads())
            times.append(elapsed);nodes.append(points)
            if baseline is None:baseline=snapshots
            for mode,data in snapshots.items():
                errors={};hasher=hashlib.sha256()
                for key in sorted(data):
                    value=np.asarray(data[key]);initial=np.asarray(baseline[mode][key])
                    if value.shape!=initial.shape or not np.all(np.isfinite(value)) or not np.allclose(value,initial,rtol=1e-12,atol=1e-14):
                        raise RuntimeError(f"Cantera numerical replay failed: {repetition} {mode} {key}")
                    errors[key]=float(np.max(np.abs(value-initial)))
                    hasher.update(key.encode());hasher.update(str(value.shape).encode());hasher.update(value.tobytes())
                replay.append(dict(repetition=repetition,mode=mode,sha256=hasher.hexdigest(),max_abs=errors))
                if repetition==0:np.savez(cantera/f"{a.case}-{mode}-first.npz",**data)
            print("cantera",a.case,repetition,sum(elapsed),elapsed,points,flush=True)
            if repetition==a.reps:
                for mode,data in snapshots.items():np.savez(cantera/f"{a.case}-{mode}-0.npz",**data)
        report["cantera_stage_seconds"]=times;report["cantera_stage_points"]=nodes
        report["cantera_repetition_thread_checks"]=thread_checks
        report["cantera_replay"]=dict(passed=True,checked_stages=len(replay),relative_tolerance=1e-12,absolute_tolerance=1e-14,records=replay)
    def run_julia():
        command=[a.julia,f"--project={a.project}",str(Path(__file__).with_suffix(".jl")),str(a.parameters),str(native),a.case,str(a.reps)]
        result=subprocess.run(command,capture_output=True,text=True,env=dict(os.environ))
        (a.output/"julia.log").write_text(result.stdout+result.stderr)
        print(result.stdout,end="",flush=True)
        if result.returncode:raise RuntimeError(f"Julia failed: {result.stderr[-3000:]}")
        data=np.load(native/"timings.npz")
        if int(data["julia_threads"][0])!=1 or int(data["blas_threads"][0])!=1:raise RuntimeError("Julia thread guard failed")
        if platform.system()=="Darwin" and int(data["accelerate_threading"][0])!=1:raise RuntimeError("Julia Accelerate thread guard failed")
        text=lambda key:bytes(data[key+"_utf8"]).decode()
        if text("kernel")!=host["kernel_release"]:raise RuntimeError("Julia and Cantera host/kernel mismatch")
        if Path(text("package_path")).resolve()!=(a.project/"src/Arrhenius.jl").resolve():raise RuntimeError("Julia loaded a different Arrhenius checkout")
        helper=a.project/"example/flames/source_flame_sequence.jl"
        if Path(text("shared_calculation_path")).resolve()!=helper.resolve() or text("shared_calculation_sha256")!=digest(helper):
            raise RuntimeError("Julia source calculation does not match the public example helper")
        report["shared_public_calculation_verified"]=True
        report["julia_stage_seconds"]=data["stage_seconds"].tolist();report["julia_stage_points"]=data["stage_points"].tolist()
        report["julia_runtime"]={key:text(key) for key in ["julia_version","kernel","machine","package_path","blas_config","loaded_libraries"]}
        report["julia_runtime"]["accelerate_threading"]=int(data["accelerate_threading"][0])
        checks=tomllib.loads(text("thread_checks_toml"))
        if len(checks)!=a.reps+2 or any(value!=1 for check in checks.values() for value in check.values()):
            raise RuntimeError("incomplete or failed Julia repetition thread checks")
        actual_sources=tomllib.loads(text("source_hashes_toml"))
        if not data["source_hashes_unchanged"][0] or actual_sources!=source_hashes:
            raise RuntimeError("Julia source bytes changed or differ from recorded checkpoint")
        report["julia_repetition_thread_checks"]=checks
        report["julia_source_bytes_verified"]=True
        report["julia_runtime"]["julia_executable_sha256"]=text("julia_executable_sha256")
        report["julia_library_hashes"]=tomllib.loads(text("library_hashes_toml"))
        expected=(a.reps+1)*(4 if a.case=="free" else 2)
        if int(data["replay_checked_stages"][0])!=expected:raise RuntimeError("incomplete Julia replay checks")
        report["julia_replay"]=dict(passed=True,checked_stages=expected,relative_tolerance=1e-12,absolute_tolerance=1e-14,
            hashes=text("replay_hashes").splitlines(),fields=text("replay_fields").splitlines(),max_abs=data["replay_max_abs"].tolist())
        report["julia_blas_library_hashes"]={path:digest(path) for path in text("loaded_libraries").splitlines()
            if any(word in Path(path).name.lower() for word in ["blas","lapack"]) and Path(path).is_file()}
    try:
        for run in ([run_ct,run_julia] if a.order=="cantera-first" else [run_julia,run_ct]):run()
        report["threads_after"]=verify_threads()
        hashes=verified_cantera_libraries(build)
        if hashes!=libraries_before:raise RuntimeError("loaded Cantera libraries changed during calculation")
        report["actual_loaded_cantera_hashes"]=hashes
        checker=Path(__file__).with_name("conservative_source_accuracy.py")
        accuracy=a.output/"accuracy.json"
        result=subprocess.run([sys.executable,str(checker),str(mechanism),str(native),str(a.references),a.case,str(accuracy)],capture_output=True,text=True)
        print(result.stdout,end="",flush=True)
        report["accuracy"]=json.loads(accuracy.read_text()) if accuracy.is_file() else dict(passed=False,error=result.stderr)
        require_unchanged(input_hashes)
        if current_source_hashes(a.project)!=source_hashes:raise RuntimeError("native source files changed during benchmark")
        if {path.name:digest(path) for path in a.references.glob(a.case+"-*.npz")}!=report["reference_hashes"]:
            raise RuntimeError("reference file set changed during benchmark")
        report["input_and_source_bytes_unchanged"]=True
        stages=4 if a.case=="free" else 2
        ct_times=checked_times(report["cantera_stage_seconds"],a.reps,stages)
        jl_times=checked_times(report["julia_stage_seconds"],a.reps,stages)
        report["cantera_seconds"]=ct_times.tolist();report["julia_seconds"]=jl_times.tolist()
        ratio=float(statistics.median(ct_times[1:])/statistics.median(jl_times[1:]))
        report["median_speed_ratio_cantera_over_julia"]=ratio
        physical=result.returncode==0 and report["accuracy"]["passed"]
        replay=report["cantera_replay"]["passed"] and report["julia_replay"]["passed"]
        report["ratio_threshold_met"]=ratio>=.95
        report["physical_gate_application"]="Last repetition of every source stage checked against refined references; strict numerical replay establishes the same outputs for all first/warm repetitions."
        report["performance_pass"]=bool(a.formal and physical and replay and ratio>=.95)
        report["passed"]=bool(physical and replay and (not a.formal or report["performance_pass"]))
    except Exception as error:
        report["failure"]=str(error)
    (a.output/"report.json").write_text(json.dumps(report,indent=2))
    print(json.dumps({key:report.get(key) for key in ["case","passed","median_speed_ratio_cantera_over_julia","failure"]}),flush=True)
    return 0 if report["passed"] else 1
if __name__=="__main__":raise SystemExit(main())

File: validation/conservative_source_bench.jl
# Complete published transport sequences; no reference states enter a solve.
using Arrhenius,LinearAlgebra,NPZ,Libdl,SHA,TOML
include("numerical_threads.jl")
thread_checks=Dict("before_first"=>benchmark_julia_thread_settings())
source_root=dirname(dirname(pathof(Arrhenius)))
function source_hashes(root)
    Dict(relpath(joinpath(folder,file),root)=>bytes2hex(sha256(read(joinpath(folder,file))))
        for (folder,_,files) in walkdir(joinpath(root,"src")) for file in files if endswith(file,".jl"))
end
initial_source_hashes=source_hashes(source_root)
parameters,output,case=ARGS[1:3]
reps=parse(Int,ARGS[4]);reps>=5 || error("at least five warm repetitions required")
Threads.nthreads()==1 && BLAS.get_num_threads()==1 || error("single-thread benchmark required")
clock_diagnostics=length(ARGS)>=5 && ARGS[5]=="clock-diagnostics"
clock_diagnostics && !(Sys.islinux() && Sys.WORD_SIZE==64) && error("clock diagnostics require 64-bit Linux")
function flame_clock_observation()
    cpu=ccall(:sched_getcpu,Cint,())
    cpu>=0 || error("sched_getcpu failed")
    return (time_ns(),ccall(:clock,Clong,())/1e6,cpu)
end
mkpath(output)
mechanism=joinpath(parameters,case=="fixed" ? "gri30.yaml" : "h2o2.yaml")
gas=CreateSolution(mechanism);data=MultiTransportData(mechanism*".multicomponent.npz",gas;mechanism)
profile=case=="fixed" ? npzread(joinpath(parameters,"fixed-profile.npz")) : nothing
shared_helper=joinpath(source_root,"example","flames","source_flame_sequence.jl")
include(shared_helper)
realpath(String(which(run_source_sequence,Tuple{typeof(gas),typeof(data),typeof(case),typeof(profile)}).file))==realpath(shared_helper) ||
    error("source calculation must come from the public example helper")
if case=="fixed"
    original_profile=source_flame_temperature_profile()
    all(profile[key]==original_profile[key] for key in ("positions","temperatures")) ||
        error("public prescribed temperature profile differs from pinned source data")
end

modes=case=="free" ? ["mass","mass-soret","multi","multi-soret"] : ["mole","multi"]
fields=["grid","T","Y","velocity","inlet_Y","state","P"]
seconds=zeros(reps+1,length(modes));points=zeros(Int,size(seconds));snapshots=nothing;first_profiles=nothing
replay_errors=zeros(reps+1,length(modes),length(fields));hashes=String[]
automatic_gc_seconds=zeros(reps+1);warmup_gc_seconds=Ref(0.0)
clock_observations=zeros(clock_diagnostics ? reps+1 : 0,4)
for repetition in 0:reps
    gc_start_ns=Base.gc_time_ns()
    clock_before=clock_diagnostics ? flame_clock_observation() : nothing
    elapsed,nodes,saved=run_source_sequence(gas,data,case,profile;save_profiles=true)
    if clock_diagnostics
        after=flame_clock_observation()
        clock_observations[repetition+1,:].=((after[1]-clock_before[1])/1e9,after[2]-clock_before[2],clock_before[3],after[3])
    end
    automatic_gc_seconds[repetition+1]=(Base.gc_time_ns()-gc_start_ns)/1e9
    thread_checks["after_repetition_"*string(repetition)]=benchmark_julia_thread_settings(;enforce=false)
    seconds[repetition+1,:].=elapsed;points[repetition+1,:].=nodes
    repetition==0 && (global first_profiles=saved)
    for (stage,mode) in enumerate(modes)
        buffer=IOBuffer()
        for (field,key) in enumerate(fields)
            current,baseline=saved[mode][key],first_profiles[mode][key]
            size(current)==size(baseline) || error("replay dimensions changed for $case $mode $key")
            all(isfinite,current) || error("nonfinite replay output")
            all(abs.(current.-baseline) .<= 1e-14 .+ 1e-12.*abs.(baseline)) || error("numerical replay failed for $case $mode $key")
            replay_errors[repetition+1,stage,field]=maximum(abs,current.-baseline)
            write(buffer,key);write(buffer,Int64.(collect(size(current))));write(buffer,vec(current))
        end
        push!(hashes,bytes2hex(SHA.sha256(take!(buffer))))
        repetition==0 && npzwrite(joinpath(output,"$case-$mode-first.npz"),saved[mode])
    end
    repetition==reps && (global snapshots=saved)
    println((;case,repetition,seconds=sum(elapsed),stages=elapsed,points=nodes));flush(stdout)
    # Exclude garbage left by the excluded warmup; automatic GC stays enabled
    # throughout all measured repetitions. No per-repetition collection.
    if repetition==0
        warmup_gc_seconds[]=@elapsed GC.gc()
    end
end
for (mode,values) in snapshots
    npzwrite(joinpath(output,"$case-$mode-0.npz"),values)
end
initial_source_hashes==source_hashes(source_root) || error("native source files changed during calculation")
tomlbytes(value)=collect(codeunits(sprint(io->TOML.print(io,value;sorted=true))))
library_hashes=Dict(realpath(path)=>bytes2hex(sha256(read(path))) for path in Libdl.dllist() if isfile(path))
npzwrite(joinpath(output,"timings.npz"),Dict("stage_seconds"=>seconds,"stage_points"=>points,
    "automatic_gc_seconds"=>automatic_gc_seconds,"warmup_gc_seconds"=>[warmup_gc_seconds[]],
    "clock_diagnostics"=>[clock_diagnostics],"clock_observations"=>clock_observations,
    "shared_calculation_path_utf8"=>collect(codeunits(realpath(shared_helper))),
    "shared_calculation_sha256_utf8"=>collect(codeunits(bytes2hex(sha256(read(shared_helper))))),
    "thread_checks_toml_utf8"=>tomlbytes(thread_checks),
    "source_hashes_toml_utf8"=>tomlbytes(initial_source_hashes),"source_hashes_unchanged"=>[true],
    "library_hashes_toml_utf8"=>tomlbytes(library_hashes),
    "julia_executable_sha256_utf8"=>collect(codeunits(bytes2hex(sha256(read(joinpath(Sys.BINDIR,"julia")))))),
    "replay_checked_stages"=>[(reps+1)*length(modes)],"replay_max_abs"=>replay_errors,
    "replay_hashes_utf8"=>collect(codeunits(join(hashes,"\n"))),"replay_fields_utf8"=>collect(codeunits(join(fields,"\n"))),
    "julia_version_utf8"=>collect(codeunits(string(VERSION))),"julia_threads"=>[Threads.nthreads()],"blas_threads"=>[BLAS.get_num_threads()],
    "accelerate_threading"=>[get(thread_checks["after_repetition_"*string(reps)],"accelerate_threading_mode",-1)],
    "kernel_utf8"=>collect(codeunits(readchomp(`uname -r`))),"machine_utf8"=>collect(codeunits(string(Sys.MACHINE))),
    "package_path_utf8"=>collect(codeunits(pathof(Arrhenius))),
    "blas_config_utf8"=>collect(codeunits(string(BLAS.get_config()))),
    "loaded_libraries_utf8"=>collect(codeunits(join(Libdl.dllist(),"\n")))))
